<img src="./logo.png" alt="Header" align="right" style="width: 100px;"/>
<center>
    <font size="+2">
    <b>Artificial Intelligence<br></b>
        <b>Project 1: 8-Puzzle</b>
    </font>
</center>


In this project you will implement the informed version of A* to solve the 8 puzzle. You will implement 2 different heuristic measures as defined in the slides and report the results. You will also develop your own heuristic measure to solve this problem. For this you will need to implement these heuristic measures in the class State. You will also have to research how to create a priority queue in python and how you can make the class `State` comparable. 

1.1 Add a variable g that will hold the cumulative backward cost

1.2 Implement the Tiles heuristic measure to compute h

1.3 Implement the Manhattan heuristic to compute h

1.4 Test both implementations and report the values for nodes visited and explored. 

1.5 Design you own heuristic measure and implement it (It has to be consistent)

In [50]:
import numpy as np
class State:
    def __init__(self,s, zero, prev=None, g=0):
        self.s=s
        self.prev=prev
        self.zero=zero
        self.hash=None
        self.g=g
        
    def __hash__(self):
        if self.hash:
            return self.hash
        s=0
        for i in range(8,-1,-1):
            s=10*s+self.s[i]
        self.hash=s
        return s
        
    #1,5,6
    #2,0,3
    #4,7,8
    
    #[1,5,6,2,0,3,4,7,8] zero=4
    def Successors(self):
        succ=[]
        s=self.s
        zero=self.zero
        g=self.g
        if (zero-1)>=0 and ((zero-1)//3==zero//3):
            tmp=s.copy()
            tmp[zero-1],tmp[zero]=s[zero],s[zero-1]
            succ.append(State(tmp,zero-1,self,g+1))
        if (zero+1<9) and ((zero+1)//3==zero//3):
            tmp=s.copy()
            tmp[zero+1],tmp[zero]=s[zero],s[zero+1]
            succ.append(State(tmp,zero+1,self,g+1))
        if (zero-3)>=0:
            tmp=s.copy()
            tmp[zero-3],tmp[zero]=s[zero],s[zero-3]
            succ.append(State(tmp,zero-3,self,g+1))
        if (zero+3)<=8:
            tmp=s.copy()
            tmp[zero+3],tmp[zero]=s[zero],s[zero+3]
            succ.append(State(tmp,zero+3,self,g+1))
        
        return succ
    
    def IsGoal(self):
        solution=[1,2,3,4,5,6,7,8,0]
        s=self.s
        for i in range(9):
            if s[i]!=solution[i]:
                return False
        return True
    
    def __eq__(self, other):
        for i in range(9):
            if self.s[i]!=other.s[i]:
                return False
        return True
    
    def __lt__(self, other):
        if self.g + self.tiles_h() < other.g + other.tiles_h():
            return True
        return False

    def tiles_h(self):
        s=self.s
        solution=[1,2,3,4,5,6,7,8,0]
        fc=0
        for i in range(9):
            if s[i]!=solution[i] and s[i]!=0:
                fc=fc+1
        return fc
    
    def manhattan_h(self):
        s=self.s
        solution=[1,2,3,4,5,6,7,8,0]
        fc=0
        for i in range(9):
            if s[i]!=0:
                f=abs(i%3-solution.index(s[i])%3)+abs(i//3-solution.index(s[i])//3)
                fc=fc+f
        return fc
    
    def __str__(self):
        
        tmp=np.array(self.s)
        return str(tmp.reshape(3,3))

In [51]:
def Solution(state):
    path=[]
    length=0
    while state:
        path.append(str(state))
        state=state.prev
        length=length+1
    return ('\n'.join(path), length)

# A*

Change the following code to implement the A* algorithm. 

In [52]:
start=State([7,2,4,5,0,6,8,3,1],4)
visited=set()
solved=False
from queue import PriorityQueue
q = PriorityQueue() 
q.put(start)
i=0
while q.qsize() > 0:
    i=i+1
    current=q.get()
    if current.IsGoal():
        path, length= Solution(current)
        print(path)
        print("Moves: "+str(length))
        solved=True
        break
    if current in visited:
        continue
    visited.add(current)
    for tmp in current.Successors():
        q.put(tmp)
    
if not solved: 
    print("No Solution")

print("Visited states: "+str(len(visited)))
print("Explored: "+str(i))

[[1 2 3]
 [4 5 6]
 [7 8 0]]
[[1 2 3]
 [4 5 0]
 [7 8 6]]
[[1 2 3]
 [4 0 5]
 [7 8 6]]
[[1 2 3]
 [0 4 5]
 [7 8 6]]
[[0 2 3]
 [1 4 5]
 [7 8 6]]
[[2 0 3]
 [1 4 5]
 [7 8 6]]
[[2 4 3]
 [1 0 5]
 [7 8 6]]
[[2 4 3]
 [0 1 5]
 [7 8 6]]
[[2 4 3]
 [7 1 5]
 [0 8 6]]
[[2 4 3]
 [7 1 5]
 [8 0 6]]
[[2 4 3]
 [7 0 5]
 [8 1 6]]
[[2 4 3]
 [7 5 0]
 [8 1 6]]
[[2 4 0]
 [7 5 3]
 [8 1 6]]
[[2 0 4]
 [7 5 3]
 [8 1 6]]
[[0 2 4]
 [7 5 3]
 [8 1 6]]
[[7 2 4]
 [0 5 3]
 [8 1 6]]
[[7 2 4]
 [5 0 3]
 [8 1 6]]
[[7 2 4]
 [5 3 0]
 [8 1 6]]
[[7 2 4]
 [5 3 6]
 [8 1 0]]
[[7 2 4]
 [5 3 6]
 [8 0 1]]
[[7 2 4]
 [5 0 6]
 [8 3 1]]
Moves: 21
Visited states: 2917
Explored: 4870


## Fill the values here

Nodes Visited (Tiles): 2917

Nodes Visited (Manhattan): 281

Nodes Visited (Yours):

Nodes Explored (Tiles): 4870

Nodes Explored (Manhattan): 412

Nodes Explored (Yours): 

Name 1 : Khalil Kassab

Name 2 : Marnie Kassar

Name 3 : Rim Chidiac (El)

We couldn't find a better heuristic measure than Manhattan.